src/percy-agent-client/dom.ts

```diff
@@ -0,0 +1,213 @@
+export interface DOMOptions {
+  enableJavaScript?: boolean,
+  domTransformation?: (dom: HTMLDocument) => void
+}
+
+/**
+ * A single class to encapsulate all DOM operations that need to be performed to
+ * capture the customer's application state.
+ *
+ */
+class DOM {
+  originalDOM: HTMLDocument
+  clonedDOM: HTMLDocument
+  options: DOMOptions
+
+  readonly defaultDoctype = '<!DOCTYPE html>'
+
+  constructor(dom: HTMLDocument, options?: DOMOptions) {
+    this.originalDOM = dom
+    this.options = options || {}
+    this.clonedDOM = this.cloneDOM()
+  }
+
+  /**
+   * Returns the final DOM string with all of the necessary transforms
+   * applied. This is the string that is passed to the API and then rendered by
+   * our API.
+   *
+   */
+  snapshotString(): string {
+    // any since the cloned DOMs type can shift
+    let dom = this.clonedDOM as any
+    const doctype = this.getDoctype()
+
+    // Sometimes you'll want to transform the DOM provided into one ready for snapshotting
+    // For example, if your test suite runs tests in an element inside a page that
+    // lists all yours tests. You'll want to "hoist" the contents of the testing container to be
+    // the full page. Using a dom transformation is how you'd achieve that.
+    if (this.options.domTransformation) {
+      try {
+        dom = this.options.domTransformation(dom)
+      } catch (error) {
+        console.error('Could not transform the dom: ', error)
+      }
+    }
+
+    return doctype + dom.outerHTML
+  }
+
+  private getDoctype(): string {
+    return this.clonedDOM.doctype
+      ? this.doctypeToString(this.clonedDOM.doctype)
+      : this.defaultDoctype
+  }
+
+  private doctypeToString(doctype: DocumentType): string {
+    const publicDeclaration = doctype.publicId
+      ? ` PUBLIC "${doctype.publicId}" `
+      : ''
+    const systemDeclaration = doctype.systemId
+      ? ` SYSTEM "${doctype.systemId}" `
+      : ''
+
+    return (
+      `<!DOCTYPE ${doctype.name}` + publicDeclaration + systemDeclaration + '>'
+    )
+  }
+
+  /**
+   * Takes the raw DOM from the snapshot, performs necessary mutations,
+   * and finally returns a deep clone of the original DOM.
+   *
+   */
+  private cloneDOM(): HTMLDocument {
+    this.mutateOriginalDOM()
+    // Any because its type changes when given to `stabilizeDOM`
+    let clone = this.originalDOM.cloneNode(true) as any
+    clone = this.stabilizeDOM(clone)
+
+    return clone
+  }
+
+  /**
+   * Serialize parts of the DOM that aren't preserved in a deep clone. Anything
+   * that is not encoded in the DOM tree needs to be accounted for here (CSSOM,
+   * input values, canvas, etc).
+   *
+   * This method should always capture these values from the original DOM and
+   * serialize into the cloned DOM. Never mutate the original DOM.
+   *
+   */
+  private stabilizeDOM(clonedDOM: HTMLDocument): HTMLElement {
+    let stabilizedDOMClone = this.serializeInputElements(clonedDOM)
```

Comment:
Just a thought here.. We don't need to pass this var around since everything that's done here mutates the cloned DOM.